drivers/hwmon/occ.c

```diff
@@ -0,0 +1,1347 @@
+	int n = attr->index;
+	struct device *dev = hwmon_dev->parent;
+	struct occ_drv_data *data = dev_get_drvdata(dev);
+	int ret = 0;
+	struct occ_sensor *sensor;
+	int val = 0;
+
+	ret = occ_update_device(dev);
+
+	if (ret != 0) {
+		dev_err(dev, "ERROR: cannot get occ sensor data: %d\n", ret);
+		return ret;
+	}
+
+	if (!data->occ_resp.data.blocks)
+		return -1;
+
+	sensor =
+		data->occ_resp.data.blocks[data->occ_resp.freq_block_id].sensor;
+	if (!sensor)
+		return -1;
+
+	dev_dbg(dev, "freq_block_id: %d, sensor: %d\n",
+		data->occ_resp.freq_block_id, n - 1);
+	val = sensor[n - 1].sensor_id;
+	dev_dbg(dev, "freq%d sensor id: %d\n", n, val);
+
+	return sprintf(buf, "%d\n", val);
+}
+
+
+static ssize_t show_occ_freq_input(struct device *hwmon_dev,
+		struct device_attribute *da, char *buf)
+{
+	struct sensor_device_attribute *attr = to_sensor_dev_attr(da);
+	int n = attr->index;
+	struct device *dev = hwmon_dev->parent;
+	struct occ_drv_data *data = dev_get_drvdata(dev);
+	int ret = 0;
+	struct occ_sensor *sensor;
+	int val = 0;
+
+	ret = occ_update_device(dev);
+
+	if (ret != 0) {
+		dev_err(dev, "ERROR: cannot get occ sensor data: %d\n", ret);
+		return ret;
+	}
+
+	if (!data->occ_resp.data.blocks)
+		return -1;
+
+	sensor =
+		data->occ_resp.data.blocks[data->occ_resp.freq_block_id].sensor;
+	if (!sensor)
+		return -1;
+
+	dev_dbg(dev, "block_id: %d, sensor: %d\n",
+		data->occ_resp.freq_block_id, n - 1);
+	val = sensor[n - 1].value;
+	dev_dbg(dev, "freq%d sensor value: %d\n", n, val);
+
+	return sprintf(buf, "%d\n", val);
+}
+
+static ssize_t show_occ_caps(struct device *hwmon_dev,
+		struct device_attribute *da, char *buf)
+{
+	struct sensor_device_attribute_2 *attr = to_sensor_dev_attr_2(da);
+	int nr = attr->nr;
+	int n = attr->index;
+	struct device *dev = hwmon_dev->parent;
+	struct occ_drv_data *data = dev_get_drvdata(dev);
+	int ret = 0;
+	struct caps_sensor *sensor;
+	int val = 0;
+
+	ret = occ_update_device(dev);
+	if (ret != 0) {
+		dev_err(dev, "ERROR: cannot get occ sensor data: %d\n", ret);
+		return ret;
+	}
+
+	dev_dbg(dev, "block_id: %d, sensor: %d, nr: %d\n",
+		data->occ_resp.caps_block_id, n - 1, nr);
+	if (!data->occ_resp.data.blocks)
+		return -1;
+
+	sensor = data->occ_resp.data.blocks[data->occ_resp.caps_block_id].caps;
+	if (!sensor)
+		return -1;
+
+	switch (nr) {
+	case 0:
+		val = sensor[n - 1].curr_powercap;
+		break;
+	case 1:
+		val = sensor[n - 1].curr_powerreading;
+		break;
+	case 2:
+		val = sensor[n - 1].norm_powercap;
+		break;
+	case 3:
+		val = sensor[n - 1].max_powercap;
+		break;
+	case 4:
+		val = sensor[n - 1].min_powercap;
+		break;
+	case 5:
+		val = sensor[n - 1].user_powerlimit;
+		break;
+	default:
+		val = 0;
+	}
+
+	dev_dbg(dev, "caps%d sensor value: %d, nr: %d\n", n, val, nr);
+
+	return sprintf(buf, "%d\n", val);
+}
+
+static struct sensor_device_attribute temp_input[] = {
+	SENSOR_ATTR(temp1_input, S_IRUGO, show_occ_temp_input, NULL, 1),
```

Comment:
There might be a really good reason for this but I have to ask, is there a reason you're starting the indexes with 1? The reason I ask is that the rest of the code is littered with 'n - 1' which is error prone.